src/algorithm.h

```diff
@@ -0,0 +1,90 @@
+#pragma once
+// Pragma once is supported on all modern compilers
+// TODO: decide whether to use include guards anyway
+
+#include <algorithm>
+
+//--------------------------------------------------------------------------------------------------
+// Container-wide versions of std algoithms.
+//--------------------------------------------------------------------------------------------------
+
+template <typename Container, typename Predicate>
```

Comment:
They're pretty thin, and reduce the noise of begin() end() everywhere when container-wide use of the standard algorithms accounts for probably 90% or more of all usage. And, admittedly and appeal to authority, Bjarne Stroustrup would agree :P.

There are a few other things in this pull request that I'm not happy with atm, so don't judge it all yet; resolving the TODO's would help me get this in shape sooner.